commit 364bb7675a793a175a4db1d815b1541b29ec1714
Author: Jessica Clarke <jrtc27@jrtc27.com>
Date:   Wed Nov 23 02:12:42 2022 +0000

    drm: Support drm-subtree drivers on FreeBSD
    
    The drm-kmod drivers use linuxkpi and end up with /dev/drm being the
    canonical path for the devices, but the drm-subtree drivers use drmkpi
    which has them appear under /dev/dri like Linux. Thus, adapt path_is_drm
    to recognise both on FreeBSD.

diff --git a/common/drm.c b/common/drm.c
index 45ed7e5..cb57a0f 100644
--- a/common/drm.c
+++ b/common/drm.c
@@ -11,6 +11,7 @@
 #define DRM_IOCTL_DROP_MASTER DRM_IO(0x1f)
 
 #define STRLEN(s) ((sizeof(s) / sizeof(s[0])) - 1)
+#define STR_HAS_PREFIX(prefix, s) (strncmp(prefix, s, STRLEN(prefix)) == 0)
 
 int drm_set_master(int fd) {
 	return ioctl(fd, DRM_IOCTL_SET_MASTER, 0);
@@ -20,17 +21,16 @@ int drm_drop_master(int fd) {
 	return ioctl(fd, DRM_IOCTL_DROP_MASTER, 0);
 }
 
-#if defined(__linux__) || defined(__NetBSD__)
+#if defined(__linux__) || defined(__NetBSD__) || defined(__FreeBSD__)
 int path_is_drm(const char *path) {
-	static const char prefix[] = "/dev/dri/";
-	static const int prefixlen = STRLEN(prefix);
-	return strncmp(prefix, path, prefixlen) == 0;
-}
-#elif defined(__FreeBSD__)
-int path_is_drm(const char *path) {
-	static const char prefix[] = "/dev/drm/";
-	static const int prefixlen = STRLEN(prefix);
-	return strncmp(prefix, path, prefixlen) == 0;
+	if (STR_HAS_PREFIX("/dev/dri/", path))
+		return 1;
+#ifdef __FreeBSD__
+	/* Some drivers have /dev/dri/X symlinked to /dev/drm/X */
+	if (STR_HAS_PREFIX("/dev/drm/", path))
+		return 1;
+#endif
+	return 0;
 }
 #else
 #error Unsupported platform
